### ISSUE TITLE
SCTransform creates many half clusters of cell types (cd4 and cd8 t-cells)

### ISSUE THREAD

**Issue by gilstel** [satijalab/seurat#5667]
I used SCtransform to normalize the data (2 treatements x 3 replicates = 18 samples, cell counts range 1650 - 3000 per sample = ~40,000 cells total)
After clustering (using 30 pcs and 1.2 resolution), several cell type markers were used to identify and label cell types.

![lymphNodePcs30res1_2namedCellTypeUmapSctPcs30Res1_2ClrByClust](https://user-images.githubusercontent.com/50887514/155875385-796fdf1a-ca5b-4bff-9d8a-21ccb32a6fd7.jpg)

Using this clustering T-cells (and B-cells) were grouped together but several satellite clusters were also created.  An phenomenon that is new to me occurred whereby half of the clusters (main and satellite) were composed of cd4 t-cells and half cd8 t-cells.

![tCellCd4](https://user-images.githubusercontent.com/50887514/155875555-f5c92678-2293-4ad8-801e-73ba52e3006c.jpg)

![tCellCd8](https://user-images.githubusercontent.com/50887514/155875563-34fbbdf4-0084-4c15-9b7f-e1097d9831a2.jpg)

This phenomenon is also evident when we subcluster the t-cells (many more satellite clusters that are include both cd4 and cd8 cells in the same cluster).

When I use the log normalized data to cluster with the same parameters (30pcs and 1.2 resolution) the phenomenon does not exist.

![tCellCd4](https://user-images.githubusercontent.com/50887514/155875666-403ccd2b-3d25-4023-86ee-7d1603f5dae3.jpg)

![tCellCd8](https://user-images.githubusercontent.com/50887514/155875670-2232e932-8462-4ced-8827-de3d2db1646c.jpg)

Did anyone encounter such a phenomenon?

Thanks,
Gil

**Comment by saketkc** [comment]
Can you post your entire code here?

**Comment by gilstel** [comment]
I wrapped the Seurat code with functions of my own so it will be problematic
Is there another solution?

**Comment by saketkc** [comment]
It might be hard to pinpoint the issue without looking at the code. But, 1) how did you run SCTransform (individually or on the entire set of samples at once) 2) Is the UMAP above on integrated space?

See [this vignette](https://satijalab.org/seurat/articles/sctransform_v2_vignette.html) for running SCTransform on case/control samples (performing integration, visualization and differential expression)

**Comment by gilstel** [comment]
The steps I ran were - 
Merged all the samples into one Seurat object
Filtered cells (according to gene count, UMI count, mito.percent and removed cells expressing hemoglobin genes)
Log normalized using NormalizeData and SCaleData(regressing out UMI count and mito.percent)
Normalize using SCTransform (regressing out UMI count and mito.percent)
FindNeighbors and FindClusters

I did not integrate the data before performing the steps 

I read the vignette carefully before performing the steps above

This exact phenomenon (different cell types located within the same cluster) has not occurred before

**Comment by saketkc** [comment]
If you are using SCTransform, `NormalizeData` and `ScaleData` are not required (this is also mentioned in the SCTransform [vignette](https://satijalab.org/seurat/articles/sctransform_vignette.html)). You should also not be regressing `nCount_RNA`, and ideally SCTransform should be run on each sample separately (to learn the sample specific noise model). You can check for batch effects by coloring the merged samples (after running SCTransform) on their batch. It appears that integration might be necessary here (you can confirm).

**Comment by gilstel** [comment]
Thanks for the info.
I read the vignette and it may be helpful to stress the points you mentioned (not to regress nCount_RNA, which is recommended using NormalizeData and ScaleData), running SCTransform on each sample individually (which I performed when trying to integrate the data to correct for batch effects, as the appropriate vignette suggests).  All the samples were labeled with an antibody in order to perform cell hashing, then processed in the parallel (same 10x Chromium plate, library construction, sequenced on the same flowcell) so there was no initial indication for an expected batch effect.
I will try to perform SCTransform on each sample separately, merge and color by condition or 10x Chromium channel to see it there is a difference.

**Comment by gilstel** [comment]
The reason I also used NormalizeData and ScaleData is because some of the plots can only this normalization.

**Comment by denvercal1234GitHub** [comment]
@saketkc - Thanks Saket for the info. Related question, if you would not mind. Thank you so much!

SCENARIO 1: If we have two cell populations of interest (X and Y) coming from the same cell type and same tissue type sequenced separately, after doing cell and gene filtering for each cell population data, should I merge these 2 cell population data first then do SCtransform or should still do SCtransform first before merging? Our goal is to identify cell subsets in X when analyzed on the background of Y cells.

SCENARIO 2: If we have sequence data from the same cell population from two different tissue types, and we want to see whether cell subset in one tissue is present in the other tissue, I will perform SCTransform on each separately, but should I integrate the two datasets or merge them? In either merging or integrating?

**Comment by gilstel** [comment]
> If you are using SCTransform, `NormalizeData` and `ScaleData` are not required (this is also mentioned in the SCTransform [vignette](https://satijalab.org/seurat/articles/sctransform_vignette.html)). You should also not be regressing `nCount_RNA`, and ideally SCTransform should be run on each sample separately (to learn the sample specific noise model). You can check for batch effects by coloring the merged samples (after running SCTransform) on their batch. It appears that integration might be necessary here (you can confirm).

I ran the following steps - 
sprt.norm.sample.list = lapply(X = sample.list, FUN = SCTransform, method = "glmGamPoi", conserve.memory = TRUE)
features = SelectIntegrationFeatures(object.list = sprt.norm.sample.list, nfeatures = 3000)

I merged the samples in the list into an object called sprt.sctransform.then.merge.sample.list

VariableFeatures(object = sprt.sctransform.then.merge.sample.list) = features
sprt.sctransform.then.merge.sample.list = RunPCA(sprt.sctransform.then.merge.sample.list, verbose = FALSE)
sprt.sctransform.then.merge.sample.list = RunUMAP(sprt.sctransform.then.merge.sample.list, reduction = "pca", dims = 1:30)

Although there were no batches in this experiment, I added a meta data column with condition (each condition was run in a separate 10x Chromium channel, to check if these created a batch effect)

gg = DimPlot(object = sprt.sctransform.then.merge.sample.list, split.by = "condition", pt.size = 0.5, ncol = 3)

![umapSplitByCondition](https://user-images.githubusercontent.com/50887514/157550056-c29fd7e0-df19-472d-8b76-56d2c45829e3.jpg)

There might be a slight "batch" effect

I continued to integrate the data as recommended (with the unmerged sct normalized sample list) - 
sprt.norm.sample.list = PrepSCTIntegration(object.list = sprt.norm.sample.list, anchor.features = features)
sprt.norm.sample.list = lapply(X = sprt.norm.sample.list, FUN = RunPCA, features = features)
ad.wt.anchors.per.sample = FindIntegrationAnchors(object.list = sprt.norm.sample.list, normalization.method = "SCT",
                                                  anchor.features = features, dims = 1:30, reduction = "rpca", k.anchor = 20)
ad.wt.combined.sct.per.sample = IntegrateData(anchorset = ad.wt.anchors.per.sample, normalization.method = "SCT", dims = 1:30)

I tried running the integratation several times on several different strong servers (e.g. 40 CPUs and 384 Gb of memory) but it could not complete

Error in subCsp_rows(x, i, drop = drop) : 
  Cholmod error 'problem too large' at file ../Core/cholmod_sparse.c, line 89

I looked in github in the seurat help pages and only found the recommendation to add conserve.memory = TRUE to the SCTransform function call

This didn't help

Any suggestions?

Obviously, I cannot estimate if all these steps solved the issue of two cell types clustering into several distinct clusters

**Comment by gilstel** [comment]
Using vst.flavor = "v2" did not help either and IntegrateData crashed

**Comment by saketkc** [comment]
Looking at the plots, I think there is little batch effect. Also, the clustering appears to be different from your original plot - have you tried annotating them? (merging seems okay here)

**Comment by gilstel** [comment]
The plot I uploaded is for data that is not clustered yet (didn't run FindNeighbors and FindClusters)
I am in process of clustering and annotating

Do you have a solution for IntegrateData crashing?
I have another dataset which I saw has a batch effect which I need to correct.  This experiment has 4 samples with ~40,000 cells and it too keeps craching.

**Comment by saketkc** [comment]
You can try reference based integration: https://satijalab.org/seurat/articles/integration_large_datasets.html

**Comment by gilstel** [comment]
I have an additional dataset whereby we experienced problems with clustering and a few cells from a big cluster labeled "cancer" and contaminate additional clusters.
![4t1pcs30res0_4namedCellTypeUmapSctPcs30Res0_4ClrByClust](https://user-images.githubusercontent.com/50887514/158188574-10f05a00-f5a2-4877-8e85-d28da9ed4c26.jpg)

Experiment desription -
The same (4 samples) 10x Genomics libraries were sequenced on two different flowcells in parallel.  Same samples were merged using cell ranger and the web report showed that the saturation of several samples was low (around 0.3).  Therefore all samples were resequenced after 3 months (due to various time constraints).  Same samples were again merged using cell ranger and the saturation level increased for most samples.

If we were to treat samples that were run on different flowcells as different sample we would have 12 (4 samples x 3 flowcells).
During the Seurat integration process same cells appearing in different samples are named differently and there would be an illusion that we have more cells than originally processed.  Is this strategy recommended or should we try to integrate the data after it is merged by cell ranger.  I should also add that BCR and TCR libraries from the same cells were also sequenced and cell ranger takes into account BCR, TCR and GEX information when merging all data or deciding to throw away cells.

Basically I am asking what is the best option for integration -
1. Treat each sample in each flowcell separately and integrate 12 samples
2. Merge data for same samples using cell ranger and then integrate the 4 samples

Your input would be much appreciated
